File: src/workflows/refine.ts
import type { z } from "zod";

const VERBS = ["shell", "open", "agent", "herdr", "run"] as const;

type StepShape = {
  shell?: string;
  open?: string;
  agent?: string;
  herdr?: string;
  run?: string;
  stdin?: string;
  prompt?: string;
  params?: Record<string, unknown>;
  wait?: "done";
  wait_for?: string;
  timeout?: number;
  close_source?: boolean;
};

export function refineStepVerbs(step: StepShape, ctx: z.core.$RefinementCtx): void {
  const verbs = VERBS.filter((v) => step[v] !== undefined);
  if (verbs.length === 0) {
    ctx.addIssue({ code: "custom", message: "step has no verb" });
    return;
  }
  if (verbs.length > 1) {
    ctx.addIssue({ code: "custom", message: `step has multiple verbs: ${verbs.join(", ")}` });
    return;
  }
  const verb = verbs[0]!;
  if (step.stdin !== undefined && verb !== "shell") {
    ctx.addIssue({ code: "custom", message: "stdin only allowed on shell", path: ["stdin"] });
  }
  if (step.prompt !== undefined && verb !== "agent") {
    ctx.addIssue({ code: "custom", message: "prompt only allowed on agent", path: ["prompt"] });
  }
  if (step.params !== undefined && verb !== "herdr") {
    ctx.addIssue({ code: "custom", message: "params only allowed on herdr", path: ["params"] });
  }
  if (step.wait !== undefined && verb !== "agent") {
    ctx.addIssue({ code: "custom", message: "wait only allowed on agent", path: ["wait"] });
  }
  if (step.wait_for !== undefined && verb !== "open") {
    ctx.addIssue({ code: "custom", message: "wait_for only allowed on open", path: ["wait_for"] });
  }
  if (step.close_source !== undefined && verb !== "agent") {
    ctx.addIssue({
      code: "custom",
      message: "close_source only allowed on agent",
      path: ["close_source"],
    });
  }
  if (step.timeout !== undefined && step.wait === undefined && step.wait_for === undefined) {
    ctx.addIssue({
      code: "custom",
      message: "timeout requires wait or wait_for",
      path: ["timeout"],
    });
  }
  if (
    verb === "run" &&
    (step.stdin !== undefined ||
      step.prompt !== undefined ||
      step.params !== undefined ||
      step.wait !== undefined ||
      step.wait_for !== undefined ||
      step.timeout !== undefined ||
      step.close_source !== undefined)
  ) {
    ctx.addIssue({ code: "custom", message: "run steps take no modifiers", path: ["run"] });
  }
}

File: rust/src/workflow/refine.rs
//! Cross-field step rules. Port of `src/workflows/refine.ts`.

use std::collections::HashSet;

/// Verb keys in `VERBS` order — drives the multi-verb message wording.
const VERBS: [&str; 5] = ["shell", "open", "agent", "herdr", "run"];

const MODIFIERS: [&str; 7] = [
    "stdin",
    "prompt",
    "params",
    "wait",
    "wait_for",
    "timeout",
    "close_source",
];

/// One refine issue: the positioned error's optional key plus its message.
pub struct RefineIssue {
    pub key: Option<&'static str>,
    pub message: String,
}

/// `refineStepVerbs`. `present` holds the keys present in the step mapping; the
/// caller runs this only when schema-level validation did not abort the step
/// (Zod skips `superRefine` after `invalid_type`/`invalid_value`/`unrecognized_keys`).
pub fn refine_step_verbs(present: &HashSet<&str>) -> Vec<RefineIssue> {
    let verbs: Vec<&str> = VERBS.into_iter().filter(|v| present.contains(v)).collect();
    if verbs.is_empty() {
        return vec![RefineIssue {
            key: None,
            message: "step has no verb".to_string(),
        }];
    }
    if verbs.len() > 1 {
        return vec![RefineIssue {
            key: None,
            message: format!("step has multiple verbs: {}", verbs.join(", ")),
        }];
    }
    let verb = verbs[0];
    let mut issues = Vec::new();
    let placements: [(&str, bool, &str); 6] = [
        ("stdin", verb == "shell", "stdin only allowed on shell"),
        ("prompt", verb == "agent", "prompt only allowed on agent"),
        ("params", verb == "herdr", "params only allowed on herdr"),
        ("wait", verb == "agent", "wait only allowed on agent"),
        ("wait_for", verb == "open", "wait_for only allowed on open"),
        (
            "close_source",
            verb == "agent",
            "close_source only allowed on agent",
        ),
    ];
    for (key, allowed, message) in placements {
        if present.contains(key) && !allowed {
            issues.push(RefineIssue {
                key: Some(key),
                message: message.to_string(),
            });
        }
    }
    if present.contains("timeout") && !present.contains("wait") && !present.contains("wait_for") {
        issues.push(RefineIssue {
            key: Some("timeout"),
            message: "timeout requires wait or wait_for".to_string(),
        });
    }
    if verb == "run" && MODIFIERS.iter().any(|k| present.contains(k)) {
        issues.push(RefineIssue {
            key: Some("run"),
            message: "run steps take no modifiers".to_string(),
        });
    }
    issues
}
